x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.UpdateForV9;
+import org.elasticsearch.ingest.EnterpriseGeoIpTask.EnterpriseGeoIpTaskParams;
+import org.elasticsearch.license.License;
+import org.elasticsearch.license.LicenseStateListener;
+import org.elasticsearch.license.LicensedFeature;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
+import org.elasticsearch.xpack.core.XPackField;
+
+import java.util.Objects;
+
+import static org.elasticsearch.ingest.EnterpriseGeoIpTask.ENTERPRISE_GEOIP_DOWNLOADER;
+
+public class EnterpriseGeoIpDownloaderLicenseListener implements LicenseStateListener, ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(EnterpriseGeoIpDownloaderLicenseListener.class);
+    // Note: This custom type is GeoIpMetadata.TYPE, but that class is not exposed to this plugin
+    static final String INGEST_GEOIP_CUSTOM_METADATA_TYPE = "ingest_geoip";
+
+    private final PersistentTasksService persistentTasksService;
+    private final ClusterService clusterService;
+    private final XPackLicenseState licenseState;
+    private static final LicensedFeature.Momentary ENTERPRISE_GEOIP_FEATURE = LicensedFeature.momentary(
+        null,
+        XPackField.ENTERPRISE_GEOIP_DOWNLOADER,
+        License.OperationMode.PLATINUM
+    );
+    private volatile boolean licenseIsValid = false;
+    private volatile boolean hasIngestGeoIpMetadata = false;
+
+    protected EnterpriseGeoIpDownloaderLicenseListener(
+        Client client,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        XPackLicenseState licenseState
+    ) {
+        this.persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
+        this.clusterService = clusterService;
+        this.licenseState = licenseState;
+    }
+
+    @UpdateForV9 // use MINUS_ONE once that means no timeout
+    private static final TimeValue MASTER_TIMEOUT = TimeValue.MAX_VALUE;
+    private volatile boolean licenseStateListenerRegistered;
+
+    public void init() {
+        listenForLicenseStateChanges();
+        clusterService.addListener(this);
+    }
+
+    void listenForLicenseStateChanges() {
+        assert licenseStateListenerRegistered == false;
+        licenseState.addListener(this);
+        licenseStateListenerRegistered = true;
+    }
+
+    @Override
+    public void licenseStateChanged() {
+        assert licenseStateListenerRegistered;
+        licenseIsValid = ENTERPRISE_GEOIP_FEATURE.checkWithoutTracking(licenseState);
+        maybeUpdateTaskState(clusterService.state());
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
```

Comment:
Why do we need to listen for cluster changed events in addition to listening to the license state changes? Is there a reason that listening only to license state changes isn't enough?

Reply:
Our goal was to keep the impact of this change minimal for users who don't use enterprise geoip databases. So we don't start the task until we see that (1) the license is valid and (2) at least one enterprise geoip database has been configured (via RestPutDatabaseConfigurationAction).
Do you think we ought to start the task regardless of whether a database is configured?

Reply:
Ahh okay, that makes sense, I could see it either direction (either the simplicity of always starting, or the lightweightness of not starting it unless necessary). I think this is fine as-is for now then.